File: src/main/java/ch/obermuhlner/astro/gradient/correction/SampleSubtraction.java
package ch.obermuhlner.astro.gradient.correction;

public interface SampleSubtraction {
  double subtract(double sample, double delta);

  default double[] subtract(double[] sample, double[] delta, double[] result) {
    if (result == null) {
      result = new double[3];
    }
    for (int i = 0; i < 3; i++) {
      result[i] = subtract(sample[i], delta[i]);
    }
    return result;
  }
}

File: src/main/java/ch/obermuhlner/astro/gradient/correction/SampleSubtraction.kt
package ch.obermuhlner.astro.gradient.correction

open interface SampleSubtraction {
    fun subtract(sample: Double, delta: Double): Double

    fun subtract(sample: DoubleArray, delta: DoubleArray, result: DoubleArray?): DoubleArray {
        var result: DoubleArray? = result
        if (result == null) {
            result = DoubleArray(3)
        }
        for (i in 0..2) {
            result[i] = subtract(sample[i], delta[i])
        }
        return result
    }
}